accounts/abi/unpack.go

```diff
@@ -34,32 +34,36 @@ var (
 
 // ReadInteger reads the integer based on its kind and returns the appropriate value
 func ReadInteger(typ Type, b []byte) interface{} {
-	switch typ.Type {
-	case uint8T:
-		return b[len(b)-1]
-	case uint16T:
-		return binary.BigEndian.Uint16(b[len(b)-2:])
-	case uint32T:
-		return binary.BigEndian.Uint32(b[len(b)-4:])
-	case uint64T:
-		return binary.BigEndian.Uint64(b[len(b)-8:])
-	case int8T:
+	if typ.T == UintTy {
```

Comment:
it seems pretty redundant with the `typ.T==UintTy` case, could you just check the `uint256` case like it was on line 57?

Reply:
Sorry, I don't know what you mean

Reply:
I mean this:

```go
func ReadInteger(typ Type, b []byte) interface{} {
	switch typ.Size {
	case 8:
		return int8(b[len(b)-1])
	case 16:
		return int16(binary.BigEndian.Uint16(b[len(b)-2:]))
	case 32:
		return int32(binary.BigEndian.Uint32(b[len(b)-4:]))
	case 64:
		return int64(binary.BigEndian.Uint64(b[len(b)-8:]))
	default:
		// the only case left for integer is int256/uint256.
		if typ.T == UintTy {
			return ret
		}
		// big.SetBytes can't tell if a number is negative or positive in itself.
		// On EVM, if the returned number > max int256, it is negative.
		// A number is > max int256 if the bit at position 255 is set.
		ret := new(big.Int).SetBytes(b)
...
}
```

So you basically drop the first if from the function, because the code is duplicated.

Reply:
But this would return only signed types, not unsigned types. 
correct would be: 
```
switch typ.Size {
	case 8:
               if typ.T == UintTy {
                return b[len(b)-1]
                } 
		return int8(b[len(b)-1])
...
```
Your fix results in
```
panic: reflect.Set: value of type int8 is not assignable to type uint8 [recovered]
        panic: reflect.Set: value of type int8 is not assignable to type uint8
```

Reply:
Right, I did compare both and somehow I didn't see the `int` in front of it. I guess that's what becoming old feels like :cry: